### PR TITLE
Fix protection logic around RunIfChanged jobs, add unit test coverage

### PR DESCRIPTION
**Pull request by fejta**
/assign @cjwagner @stevekuznetsov

### CODE REVIEW

File: prow/cmd/branchprotector/protect_test.go [hunk 1]
```diff
@@ -27,6 +27,103 @@ import (
...
+				{
+					Context: "optional",
+					RunAfterSuccess: []config.Presubmit{
+						{
+							Context: "no thanks",
+						},
+					},
+				},
+				{
+					Context:    "hidden-grandpa",
```
Comment (stevekuznetsov) [inline prow/cmd/branchprotector/protect_test.go]: 😆

File: prow/cmd/branchprotector/protect_test.go [hunk 2]
```diff
@@ -27,6 +27,103 @@ import (
 	"k8s.io/test-infra/prow/github"
 )
 
+func TestJobRequirements(t *testing.T) {
```
Comment (stevekuznetsov) [inline prow/cmd/branchprotector/protect_test.go]: Might be easier (for me, anyway) to understand if `AlwaysRun` and `SkipReport` were always explicitly called out -- hard to keep track and brain-default them.

File: prow/cmd/branchprotector/protect_test.go (hunk 1 above)
Comment (fejta) [reply prow/cmd/branchprotector/protect_test.go]: Ack

File: prow/cmd/branchprotector/protect_test.go (hunk 2 above)
Comment (fejta) [reply prow/cmd/branchprotector/protect_test.go]: Done

**Pull request merged**